Tasks/AzureMonitorAlerts/azurerestutility.ts

```diff
@@ -0,0 +1,265 @@
+        .then((response: IRestResponse<any>) => {
+            if(response.statusCode === 200) {
+                deferred.resolve(response.result["value"][0]);
+            }
+            else {
+                tl.debug(`Getting AzureRm resource '${resourceName}' failed. Status Code: ${response.statusCode}. Response: ${JSON.stringify(response.result, null, 2)}`);
+                deferred.reject(response.statusCode);
+            }
+
+        }, (error) => {
+            deferred.reject(error);
+        });
+
+    return deferred.promise;
+}
+
+export async function getAzureRmAlertRule(
+    endpoint, 
+    resourceGroupName: string, 
+    alertRuleName: string) : Promise<any> {
+
+    let deferred: Q.Deferred<any> = Q.defer<any>();
+    let accessToken: string = await getAuthorizationToken(endpoint);
+
+    tl.debug(`Getting AzureRm alert rule - '${alertRuleName}' in resource group '${resourceGroupName}'`);
+
+    let apiVersion: string = "2016-03-01";
+    let restUrl: string = `${endpoint.url}subscriptions/${endpoint.subscriptionId}/resourceGroups/${resourceGroupName}/providers/Microsoft.insights/alertrules/${alertRuleName}?api-version=${apiVersion}`;
+    let restOptions: IRequestOptions = {
+        additionalHeaders: {
+            "Authorization": "Bearer " + accessToken
+        }
+    };
+
+    restObj.get(restUrl, restOptions)
+        .then((response: IRestResponse<any>) => {
+            if(response.statusCode === 200) {
+                deferred.resolve(response.result);
+            }
+            else {
+                tl.debug(`Getting AzureRm alert '${alertRuleName}' failed. Status Code: ${response.statusCode}. Response: ${JSON.stringify(response.result, null, 2)}`);
+                deferred.reject(response.statusCode);
+            }
+
+        }, (error) => {
+            deferred.reject(error);
+        });
+
+    return deferred.promise;
+}
+
+async function getRequestBodyForAddingAlertRule(
+    endpoint, 
+    resourceGroupName: string, 
+    resourceUri: string, 
+    rule: IAzureMetricAlertRule, 
+    notifyServiceOwners: boolean, 
+    notifyEmails: string) : Promise<any> {
+
+    let existingAlertRule: any;
+    let notifyViaEmails: boolean = notifyServiceOwners || !!notifyEmails;
+
+    try {
+        console.log(tl.loc("AlertRuleCheck", rule.alertName, resourceGroupName));
+       
+        existingAlertRule = await getAzureRmAlertRule(endpoint, resourceGroupName, rule.alertName);
+
+        let existingAlertRuleTargetResourceUri: string = existingAlertRule["properties"].condition.dataSource.resourceUri;
+        if(existingAlertRuleTargetResourceUri && existingAlertRuleTargetResourceUri.toLowerCase() !== resourceUri.toLowerCase()) {
+            return Q.reject(tl.loc("AlertRuleTargetResourceIdMismatchError", rule.alertName, existingAlertRuleTargetResourceUri));
+        }
+
+        console.log(tl.loc("AlertRuleExists", rule.alertName, resourceGroupName));
+    }
+    catch (error) {
```

Comment:
error will never be 404. In case of 404, rest client resolves with null object. null check on the response object. In general also, even if 404 was being thrown, this check would have failed. Error.message would have been 404.

Reply:
In case of 404 the response object is resolved as { statusCode: 404, result: undefined } . https://github.com/Microsoft/typed-rest-client/blob/master/lib/RestClient.ts (line 187)

Reply:
error.message will be 404. I think the error that you catch is a error object,  and not the statuscode itself. also, if that is the case. the comparison should be a string comparison and not a numerical comparison

Reply:
JS allows you to throw custom error objects and in such cases the custom error object is passed as it is to catch block. Here i am throwing a number, hence numerical comparison works just fine.